First draft

johngibb opened the pull request.

johngibb committed:
Implement fetcher

jeffreylo commented on fetcher.go:
```diff
@@ -0,0 +1,173 @@
+		wait:      wait,
+		fetch:     fetch,
+		batch:     newBatch(),
+		cache:     lru.New(10),
```
Any reason for the `10`? Would it be useful for callers to configure this?

dcowgill replied:
No doubt.

And then `New` would, perhaps, have too many parameters. Could go with an options/configuration struct.

johngibb replied:
yep, just an oversight. definitely needs to be configurable.

jeffreylo commented on fetcher.go:
```diff
@@ -0,0 +1,173 @@
+	f.muCache.Unlock()
+}
+
+func (f *Fetcher) getFromCache(ids []int64) (cached map[int64]interface{}, uncached []int64) {
```
nit: any point in preallocating the uncached slice with the length of `ids`?

johngibb replied:
I went back and forth on this. Really, the appropriate choice depends on the cache hit rate; if most requests were fully cached, you'd want this as-is. If most requests were uncached, you'd want to go with your suggestion.

Given I don't know this up front, I think it's safer to leave as-is and hope for cache hits.

jeffreylo commented on fetcher_test.go:
```diff
@@ -0,0 +1,102 @@
+)
+
+func TestDataLoader(t *testing.T) {
+	if testing.Short() {
```
good call

jeffreylo commented on lru/cache.go:
```diff
@@ -0,0 +1,66 @@
+package lru
```
in memoriam 🍁

jeffreylo commented on fetcher_test.go:
```diff
@@ -0,0 +1,102 @@
+package fetcher_test
```
Perhaps an additional test for error handling too? Should the fetcher also handle panics in the `fetch`? 

“tests are affidavits”; in other words, the tests are ideally exhaustive of use cases for the library and introduce users to what’s possible

johngibb replied:
done, feel free to review again. TLDR, panics will be caught, returned as an error of type `fetcher.PanicError`, and printing them includes the stack trace.

jeffreylo commented on lru/cache_test.go:
```diff
@@ -0,0 +1,47 @@
+		_ = c.Get(1)  // 1 is now most recently used
+		c.Put(3, "3") // should evict 2, not 1
+
+		// Cache should only have two entries: a and c.
```
```suggestion
		// Cache should only have two entries: 1 and 3.
```

jeffreylo commented on fetcher.go:
```diff
@@ -0,0 +1,173 @@
+		if val := cached[id]; val != nil {
+			results[i] = val
+		} else if batch != nil {
+			results[i] = batch.results[id]
```
Should this also handle a missing `id` in `batch.results`?

johngibb replied:
this is solved by using maps as the return values.

dcowgill commented on fetcher.go:
```diff
@@ -0,0 +1,173 @@
+
+		// Reassemble the id set into a slice of ids.
+		ids := make([]int64, 0, len(batch.ids))
+		for id, _ := range batch.ids {
```
nit: no need for the `, _` here.

dcowgill commented on fetcher.go:
```diff
@@ -0,0 +1,173 @@
+	wait      time.Duration                        // max time to wait for batch to fill up
+	batchSize int                                  // once batch exceeds this size, it's always run
+	fetch     func([]int64) ([]interface{}, error) // fetch function
+	cacheTTL  time.Duration                        // cache TTL; zero disables caching.
```
I'd consider either treating cacheTTL==0 as no expiry or simplifying the code path for disabling the cache. Right now you have to test that cacheTTL==0 in more than one spot, and it even goes through the "batcher" code path. It _might_ make sense to change Fetcher.Get to this:

```go
func (f *Fetcher) Get(ctx context.Context, ids []int64) ([]interface{}, error) {
    if f.cacheTTL == 0 {
        return f.fetch(ids) // i.e. bypass the entire caching code path
    }
    // as before...
}
```

johngibb replied:
This would bypass both caching _and_ batching though, right? I'd still want batching to occur when caching is disabled (basically falling back to the traditional [dataloader](github.com/facebook/dataloader) pattern).

dcowgill commented on fetcher.go:
```diff
@@ -0,0 +1,173 @@
+		}
+
+		// Perform the fetch.
+		results, err := f.fetch(ids)
```
Is it ok that len(ids) might be significantly greater than the desired batch size, possibly even by several multiples? (Unlikely scenario, maybe, but many goroutines might separately populate a single batch with disjoint sets of IDs.)

Or is it the responsibility of the "fetch" function to handle that case?

johngibb replied:
I thought about that, and reasoned that it's fine as-is (but probably should be documented as a caveat).

Rationale: the `batchSize` should be larger than the typical size of `ids` being requested in a single call to `Get`, or else the batching is ineffectual. Given that, a batch in a single fetch can only ever contain `batchSize + len(mostRecentRequest.ids)`, which _should_ mean it's between `[batchSize, 2*batchSize]`. For our use case, of fetching ~500 restaurants on average, or maybe ~1500 in worst case, I'd set a batch size of 1000 and be fine with occasional queries pulling 1500 (since, that's what's happening now anyway).

If it was simple to enforce the batch size, I'd be open to it, but I think it'd add a lot of complexity (since now a request may be waiting on multiple batches, and potentially getting back multiple errors).

What do you think?

johngibb replied:
re:
> possibly even by several multiples

it couldn't exceed `batchSize` by several multiples unless a single call to `Get` had a massive list of ids, right? or am I missing something?

I think the [code that executes the current batch immediately as soon as it gets oversized](https://github.com/johngibb/fetcher/pull/1/files#diff-b21be20702219cd7ad6a3288b7655a84R68) ensures that only one request is allowed to overflow the batch size; there's no possibility that several goroutines could push it over by their aggregate size of requests.

dcowgill replied:
You're no longer in the critical section at that point, so I believe another goroutine could call Fetcher.submit, acquire f.muBatch, and add more keys to f.batch.

dcowgill replied:
That said, I do think it's unlikely that the query would _greatly_ exceed the target batch size, since the submitted sets of ID are probably going to have a lot of overlap, and it doesn't look like much is happening between releasing the mutex and reacquiring it inside of the sync.Once

dcowgill replied:
BTW, I wouldn't implement the cap on batch size by creating multiple batches and waiting on multiple channels. I would just call the "fetch" function multiple times and compile the results (bailing out and discarded all results should it return an error).

dcowgill reviewed (approved):
I would probably add a test (which would not run by default, or perhaps even a separate command) that spawns a lot of goroutines which fetch random batches of IDs (using a fetch function w/ a stochastic wait time) and verifies the results. Basically, I don't trust my own concurrent code unless I thoroughly exercise it.

johngibb commented:
@dcowgill @jeffreylo: thanks for all the help! I've (hopefully) addressed all your feedback, and have gotten this to a point I'm happy with, so I'm going to merge it so I can begin to use it. Feel free to comment if you've got any other feedback and I'll address in a future commit!
